docs/decisions/01-system-design-guidance.md

```diff
@@ -0,0 +1,48 @@
+﻿# System Design Guidance
+
+Priorities for the project, and some guidance on how to balance those priorities when making high level architectural decisions.
+
+## Status
+
+- [ ] Decided (proposal was agreed and adopted)
+- [ ] Decided Against (agreed not to adopt)
+- [ ] Deferred (no decision, but should be revisited later)
+- [ ] Superseded (later events or decisions make this no longer relevant)
+
+# Decision
+
+This is about guidance, so here's the guidance.
+
+1. Prefer multiple providers. There will be system components that can be rented from managed service providers (you know, clouds). It's better to choose technologies that are available in multiple cloud providers. In the future, it would be even better to support multiple services in a category (MySQL AND PostreSQL, for example). That's not feasible now, but try not to design out the option of doing it later. 
+2. Be self-hostable. Don't choose services that are only available in the cloud (like snowflake or dynamodb), or that are difficult to administer (like kafka). This goes double if it's only available from a single provider.
+3. Plan for scale-up. Networks tend to grow over time. It should be a reasonable amount of work for a volunteer admin to add capacity as the need arises.
+
+
+## Impact
+
+Now
+* DB - Postgres
+* Object Storage - Any S3 compatible
+
+Future
+* Message Queue - Something internal (maybe subscriber pattern?), then RabbitMQ (if/when we need it)
+* Orchestration - Something internal (threads, processes), then maybe Nomad?
+
+## Context
+
+Basically, this came up when I was considering which database engine should serve as the core database for the project. There are a number of goals for Letterbook, and I want to write them down, for the future.
+
+### Optimize for the operators
+
+There is some some threshold of features that would make Letterbook viable for users. Beyond that, we should favor the needs of instance administrators, operations teams, and moderators. That group is likely to overwhelmingly be hobbyists and volunteers. They will often have shoestring budgets. They will probably live all over the world, speak numerous languages, and work in various time zones. In particular, I'm thinking of the experience that the https://hachyderm.io team had in setting up and scaling up their mastodon instance. They had a team of frankly world class SREs, and they still struggled with it continuously for weeks. Many instances had to close registrations during the twitter migration because they couldn't operationally support the added demand. You shouldn't have to have a team of senior engineers to run a community chat service.
```

Comment:
How do you see this impacting Telemetry and Observability? Adding that could make maintenance and upkeep simpler, *but* can lead to some pretty substantial costs for storing and cataloging all of that telemetry info which might conflict with the next goal.

Will we need a system for being able to toggle features on/off easily that Admin's can use?

Reply:
Yes, letterbook should produce some telemetry. I'm a fan of OpenTelemetry. When we get to that point, I'm imagining a guide or template or something that includes an OTel collector, with configs to make it easy to adjust sampling rates to keep that cost down.

Reply:
Oh, for feature flags/feature toggles/something like that: maybe? I think there's a lot of value in those systems, but I normally do those things in the context of a service I'm operating. I'm not sure how I would go about it in this case.

Reply:
One thing I'm wondering here is how much can we defer to plugins for feature support. If the core of this system is just an MVP with basic functionality and everything else is left up to plugins, this cost and optimization knob might be easier to turn (thinking sort of the WordPress model here). I'm not familiar enough yet to know exactly what that would look like, but I could see it enabling a lot of different useful things and allow admins to customize their servers to their needs. That approach might encompass the idea of feature flags to and make them redundant. This might come with some up-front complexity in setup, but it'd mean you could choose from say an elastic powered search feature or just an indexed hashtag powered one instead of us making that decision for you.

Reply:
That's an interesting idea. I've seen plugin driven systems go very badly before. Or, once. But clearly they can work. I think that's probably a question of how to support contributors. If this takes off like a rocket and there's tons of people looking to build things, then plug-ins look really appealing. If it's more gradual, or people just want to use letterbook, not build it, then my thinking is they're more trouble than they're worth.

Reply:
Maybe it's an idea worth revisiting after an MVP. I imagine we'll learn a lot about what works and doesn't work building that.